File: sdks/rust/benches/agentic_receipt_bench.rs
//! Micro-benchmark: receipt/lineage emission overhead (ADR-0028 §D7-§D9,
//! issue #56 / PR #84) — the canonicalization + digest + parse path that
//! runs on every Meta LLM response carrying a `cognitum_receipt` field
//! (ADR-0024b §D3), and on every locally-constructed `ExecutionReceipt`.
//!
//! Three real, already-shipped functions are exercised:
//!   1. `meta_llm::parse_meta_llm_receipt` — decodes the wire
//!      `cognitum_receipt` payload into a typed `MetaLlmReceipt` on every
//!      nonstream/stream response (`meta_llm::types::receipt`'s doc
//!      comment).
//!   2. `agentic::build_execution_receipt` — constructs an
//!      `ExecutionReceipt`, which internally canonicalizes (JSON, sorted
//!      keys, JS-compatible number formatting) and SHA-256-digests the
//!      signable payload.
//!   3. `agentic::verify_execution_receipt` — re-canonicalizes and
//!      re-digests the receipt to verify it (shape + digest levels here;
//!      the `cryptographic` HMAC-signature branch is exercised by the
//!      SDK's own test suite, not this bench, to avoid adding `hmac`/`sha2`
//!      as bench-only dev-dependencies when they're already fully covered
//!      by `agentic::receipt_verification`'s own unit tests).
//!
//! The fixture receipt shape matches
//! `sdks/fixtures/receipt-canonicalization/execution-receipt-v1.json` (the
//! cross-SDK canonicalization conformance fixture from PR #84's review
//! fix), so the input size/shape here is representative of a real
//! receipt, not a toy one.
//!
//! Target (engineering estimate, NOT ADR-mandated — no ADR cites a
//! receipt-canonicalization latency number): p50 < 50 µs per operation.
//! This is a synchronous, allocation-heavy but otherwise CPU-only path (JSON
//! serialize + string sort + SHA-256 over a ~600-800 byte canonical
//! payload) — no I/O, so it should be at least an order of magnitude
//! cheaper than the seed client's ADR-0005 <1ms-p50 network-overhead
//! budget.
//!
//! Run it as an example:
//!
//! ```bash
//! cargo run --release --features meta-llm --example agentic_receipt_bench
//! # or, registered as a bench target:
//! cargo bench --features meta-llm --bench agentic_receipt_bench
//! ```

#![cfg(feature = "meta-llm")]

use std::time::{Duration, Instant};

use cognitum_one::agentic::{
    build_execution_receipt, canonical_json, sha256_hex, verify_execution_receipt,
    BuildExecutionReceiptInput, CostFinality, CostObservation, VerificationLevel,
    VerifyReceiptOptions,
};
use cognitum_one::meta_llm::parse_meta_llm_receipt;
use serde_json::json;

const ITERS: usize = 20_000;

fn measure<F: FnMut()>(label: &str, iters: usize, mut f: F) -> Duration {
    for _ in 0..200 {
        f();
    }
    let mut samples = Vec::with_capacity(iters);
    for _ in 0..iters {
        let t0 = Instant::now();
        f();
        samples.push(t0.elapsed());
    }
    samples.sort();
    let p50 = samples[iters / 2];
    let p95 = samples[(iters * 95) / 100];
    let mean: Duration = samples.iter().sum::<Duration>() / iters as u32;
    println!(
        "{label:55}  mean={:>9.3}µs  p50={:>9.3}µs  p95={:>9.3}µs",
        mean.as_secs_f64() * 1_000_000.0,
        p50.as_secs_f64() * 1_000_000.0,
        p95.as_secs_f64() * 1_000_000.0,
    );
    p50
}

/// Wire `cognitum_receipt` payload — shape/size representative of a real
/// Meta LLM chat-completions response receipt (routing, price, cache,
/// safety, breaker-count evidence).
fn wire_receipt_payload() -> serde_json::Value {
    json!({
        "request_id": "req_bench_0001",
        "resolved_tier": "large",
        "resolved_model": "meta-llm-large",
        "escalated": false,
        "cap_degraded": false,
        "routing_reason": "primary_healthy",
        "price": {"amount": "12.34", "currency": "USD"},
        "cache_result": "miss",
        "cache_savings": {"amount": "0.00", "currency": "USD"},
        "fallback_used": false,
        "breaker_counts": {"primary": 0, "fallback": 0},
        "sub_tenant_id": "tenant-bench-0001",
        "safety_summary": {
            "mode": "warn",
            "detector_classes": ["pii", "secrets"],
            "blocked": false
        },
        "usage": {
            "prompt_tokens": 128,
            "completion_tokens": 64,
            "total_tokens": 192,
            "cache_hit_ratio": 0.0
        },
        "costs": [
            {"source": "openrouter", "amount": 100.0, "currency": "USD", "finality": "invoiced"},
            {"source": "meta-llm", "amount": 12.34, "currency": "USD", "finality": "estimate"}
        ]
    })
}

/// Matches `sdks/fixtures/receipt-canonicalization/execution-receipt-v1.json`'s
/// `logicalReceipt` shape/size (issue #56 / PR #84 cross-SDK conformance
/// fixture).
fn build_receipt_input() -> BuildExecutionReceiptInput<'static> {
    BuildExecutionReceiptInput {
        receipt_id: "rcpt_bench_0001".to_owned(),
        product: "meta-llm".to_owned(),
        contract_version: "1.0.0".to_owned(),
        request_id: "req_bench_abc123".to_owned(),
        operation_id: Some("op_bench_xyz789".to_owned()),
        tenant_hash: Some("th_bench_deadbeef".to_owned()),
        started_at: "2026-01-01T00:00:00.000Z".to_owned(),
        completed_at: Some("2026-01-01T00:00:05.250Z".to_owned()),
        usage: Some(json!({
            "prompt_tokens": 128,
            "completion_tokens": 64,
            "cacheHitRatio": 0.5
        })),
        costs: vec![
            CostObservation {
                source: "openrouter".to_owned(),
                amount: 100.0,
                currency: "USD".to_owned(),
                finality: CostFinality::Invoiced,
            },
            CostObservation {
                source: "meta-llm".to_owned(),
                amount: 12.34,
                currency: "USD".to_owned(),
                finality: CostFinality::Estimate,
            },
        ],
        outcome: "success".to_owned(),
        artifact_digests: Some(vec![
            "sha256:aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa".to_owned(),
            "sha256:bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb".to_owned(),
        ]),
        lineage_root: Some(
            "sha256:cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc".to_owned(),
        ),
        issuer: Some("cognitum-one/meta-llm".to_owned()),
        key_id: Some("key-2026-01".to_owned()),
        sign: None,
        now: None,
    }
}

fn main() {
    println!("Receipt/lineage emission overhead — {ITERS} iterations each\n");

    // 1. Wire-receipt parse (runs on every Meta LLM response carrying a
    // `cognitum_receipt` field).
    let raw_payload = wire_receipt_payload();
    measure("parse_meta_llm_receipt", ITERS, || {
        let receipt = parse_meta_llm_receipt(&raw_payload);
        assert!(receipt.is_some());
    });

    // 2. build_execution_receipt — construction + canonicalization + digest
    // (no signer configured; see module doc comment).
    measure("build_execution_receipt", ITERS, || {
        let receipt = build_execution_receipt(build_receipt_input());
        assert_eq!(receipt.verification.level, VerificationLevel::Shape);
    });

    // 3. verify_execution_receipt — re-canonicalize + re-digest an
    // already-built receipt (the read-side counterpart of #2).
    let sample_receipt = build_execution_receipt(build_receipt_input());
    measure("verify_execution_receipt (shape+digest)", ITERS, || {
        let result = verify_execution_receipt(
            &sample_receipt,
            &VerifyReceiptOptions {
                min_level: VerificationLevel::Digest,
                expected_digest: None,
                resolve_key: None,
                check_anchor: None,
                now: None,
            },
        );
        assert!(result.subject_digest.is_some());
    });

    // 4. canonical_json + sha256_hex alone (the shared primitive both (2)
    // and (3) call internally) — isolates the pure canonicalization/digest
    // cost from receipt-shape validation overhead.
    let value = serde_json::to_value(&sample_receipt).expect("receipt should serialize");
    measure("canonical_json + sha256_hex (primitive)", ITERS, || {
        let bytes = canonical_json(&value);
        let digest = sha256_hex(&bytes);
        assert_eq!(digest.len(), 64);
    });

    println!("\nAll four operations are expected to clear p50 < 50µs (engineering target, not ADR-mandated).");
}

File: sdks/node/bench/agentic-receipt-bench.ts
/**
 * Micro-benchmark: receipt/lineage emission overhead (ADR-0028 §D7-§D9,
 * issue #56 / PR #84) — the canonicalization + digest + parse path that
 * runs on every Meta LLM response carrying a `cognitum_receipt` field
 * (ADR-0024b §D3), and on every locally-constructed `ExecutionReceipt`.
 *
 * Three real, already-shipped functions are exercised:
 *   1. `parseMetaLlmReceipt` — decodes the wire `cognitum_receipt` payload
 *      into a typed `MetaLlmReceipt` on every nonstream/stream response
 *      (`../src/meta-llm/types/receipt.ts`'s doc comment).
 *   2. `buildExecutionReceipt` — constructs an `ExecutionReceipt`, which
 *      internally canonicalizes (JSON, sorted keys) and SHA-256-digests
 *      the signable payload.
 *   3. `verifyExecutionReceipt` — re-canonicalizes and re-digests the
 *      receipt to verify it (shape + digest levels here; the
 *      `cryptographic` HMAC-signature branch is exercised by the SDK's
 *      own unit tests, not this bench, to keep the bench dependency-free
 *      of any signer wiring).
 *
 * The fixture receipt shape matches
 * `sdks/fixtures/receipt-canonicalization/execution-receipt-v1.json` (the
 * cross-SDK canonicalization conformance fixture from PR #84's review
 * fix), so the input size/shape here is representative of a real receipt.
 *
 * Target (engineering estimate, NOT ADR-mandated — no ADR cites a
 * receipt-canonicalization latency number): p50 < 200 µs per operation
 * (looser than the Rust bench's 50µs target — V8 object allocation/GC and
 * `JSON.stringify`'s recursive key-sort walk cost more per call than the
 * Rust equivalent at this scale). No I/O — should be well under the seed
 * client's ADR-0005 <1ms-p50 network-overhead budget either way.
 *
 * Usage:
 *   npx tsx bench/agentic-receipt-bench.ts
 *   node --experimental-strip-types bench/agentic-receipt-bench.ts
 */

import {
  buildExecutionReceipt,
  canonicalJson,
  sha256Hex,
  verifyExecutionReceipt,
  type BuildExecutionReceiptInput,
} from "../dist/agentic/index.js";
import { parseMetaLlmReceipt } from "../dist/meta-llm/index.js";

const ITERS = 20_000;

function measure(label: string, iters: number, fn: () => void): number {
  for (let i = 0; i < 200; i += 1) fn();

  const samples = new Array<number>(iters);
  for (let i = 0; i < iters; i += 1) {
    const t0 = performance.now();
    fn();
    samples[i] = performance.now() - t0;
  }
  samples.sort((a, b) => a - b);
  const p50 = samples[Math.floor(iters * 0.5)];
  const p95 = samples[Math.floor(iters * 0.95)];
  const mean = samples.reduce((a, b) => a + b, 0) / iters;
  console.log(
    `${label.padEnd(55)}  mean=${(mean * 1000).toFixed(3)}µs  p50=${(p50 * 1000).toFixed(3)}µs  p95=${(p95 * 1000).toFixed(3)}µs`,
  );
  return p50;
}

/** Wire `cognitum_receipt` payload — shape/size representative of a real Meta LLM response receipt. */
function wireReceiptPayload(): unknown {
  return {
    request_id: "req_bench_0001",
    resolved_tier: "large",
    resolved_model: "meta-llm-large",
    escalated: false,
    cap_degraded: false,
    routing_reason: "primary_healthy",
    price: { amount: "12.34", currency: "USD" },
    cache_result: "miss",
    cache_savings: { amount: "0.00", currency: "USD" },
    fallback_used: false,
    breaker_counts: { primary: 0, fallback: 0 },
    sub_tenant_id: "tenant-bench-0001",
    safety_summary: { mode: "warn", detector_classes: ["pii", "secrets"], blocked: false },
    usage: { prompt_tokens: 128, completion_tokens: 64, total_tokens: 192, cache_hit_ratio: 0.0 },
    costs: [
      { source: "openrouter", amount: 100.0, currency: "USD", finality: "invoiced" },
      { source: "meta-llm", amount: 12.34, currency: "USD", finality: "estimate" },
    ],
  };
}

/** Matches `sdks/fixtures/receipt-canonicalization/execution-receipt-v1.json`'s `logicalReceipt` shape/size. */
function buildReceiptInput(): BuildExecutionReceiptInput {
  return {
    receiptId: "rcpt_bench_0001",
    product: "meta-llm",
    contractVersion: "1.0.0",
    requestId: "req_bench_abc123",
    operationId: "op_bench_xyz789",
    tenantHash: "th_bench_deadbeef",
    startedAt: "2026-01-01T00:00:00.000Z",
    completedAt: "2026-01-01T00:00:05.250Z",
    usage: { prompt_tokens: 128, completion_tokens: 64, cacheHitRatio: 0.5 },
    costs: [
      { source: "openrouter", amount: 100, currency: "USD", finality: "invoiced" },
      { source: "meta-llm", amount: 12.34, currency: "USD", finality: "estimate" },
    ],
    outcome: "success",
    artifactDigests: [
      "sha256:aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
      "sha256:bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb",
    ],
    lineageRoot: "sha256:cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc",
    issuer: "cognitum-one/meta-llm",
    keyId: "key-2026-01",
  };
}

function main() {
  console.log(`Receipt/lineage emission overhead — ${ITERS} iterations each\n`);

  // 1. Wire-receipt parse (runs on every Meta LLM response carrying a `cognitum_receipt` field).
  const rawPayload = wireReceiptPayload();
  const p50Parse = measure("parseMetaLlmReceipt", ITERS, () => {
    const receipt = parseMetaLlmReceipt(rawPayload);
    if (!receipt) throw new Error("expected receipt to parse");
  });

  // 2. buildExecutionReceipt — construction + canonicalization + digest (no signer configured).
  const p50Build = measure("buildExecutionReceipt", ITERS, () => {
    const receipt = buildExecutionReceipt(buildReceiptInput());
    if (receipt.verification.level !== "shape") throw new Error("expected shape-level receipt");
  });

  // 3. verifyExecutionReceipt — re-canonicalize + re-digest an already-built receipt.
  const sampleReceipt = buildExecutionReceipt(buildReceiptInput());
  const p50Verify = measure("verifyExecutionReceipt (shape+digest)", ITERS, () => {
    const result = verifyExecutionReceipt(sampleReceipt, { minLevel: "digest" });
    if (!result.subjectDigest) throw new Error("expected subjectDigest to be computed");
  });

  // 4. canonicalJson + sha256Hex alone (the shared primitive both (2) and (3) call internally).
  const p50Primitive = measure("canonicalJson + sha256Hex (primitive)", ITERS, () => {
    const bytes = canonicalJson(sampleReceipt);
    const digest = sha256Hex(bytes);
    if (digest.length !== 64) throw new Error("expected 64-char hex digest");
  });

  const target = 0.2; // ms == 200µs
  const allPass = [p50Parse, p50Build, p50Verify, p50Primitive].every((p50) => p50 < target);
  console.log(
    `\n${allPass ? "PASS" : "WARN"}: all four operations ${allPass ? "clear" : "do NOT all clear"} p50 < 200µs (engineering target, not ADR-mandated).`,
  );
}

main();
